File: Data Structure II/linked_list/linked_list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def insert(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
        else:
            temp = self.head
            while temp.next:
                temp = temp.next
            temp.next = new_node

    def remove(self, data):
        temp = self.head
        if not temp:
            return "List is empty."
        
        if temp.data == data:
            self.head = temp.next
            return f"Removed {data}"

        while temp.next and temp.next.data != data:
            temp = temp.next

        if temp.next:
            temp.next = temp.next.next
            return f"Removed {data}"
        return "Number not found."

    def display(self):
        elements = []
        temp = self.head
        while temp:
            elements.append(temp.data)
            temp = temp.next
        return elements if elements else "List is empty."

    def contains(self, data):
        temp = self.head
        while temp:
            if temp.data == data:
                return True
            temp = temp.next
        return False

def menu():
    ll = LinkedList()
    while True:
        print("\nLinked List Menu:")
        print("1. Insert")
        print("2. Remove")
        print("3. Display List")
        print("4. Check if Number Exists")
        print("5. Exit")
        choice = input("Enter choice: ")

        if choice == "1":
            num = int(input("Enter number to insert: "))
            ll.insert(num)
        elif choice == "2":
            num = int(input("Enter number to remove: "))
            print(ll.remove(num))
        elif choice == "3":
            print("List:", ll.display())
        elif choice == "4":
            num = int(input("Enter number to check: "))
            print("Exists" if ll.contains(num) else "Does not exist")
        elif choice == "5":
            break
        else:
            print("Invalid choice, try again.")

if __name__ == "__main__":
    menu()

File: Data Structure II/linked_list/linked_list.rb
class Node
  attr_accessor :data, :next

  def initialize(data)
    @data = data
    @next = nil
  end
end

class LinkedList
  def initialize
    @head = nil
  end

  def insert(data)
    new_node = Node.new(data)
    if @head.nil?
      @head = new_node
    else
      temp = @head
      temp = temp.next while temp.next
      temp.next = new_node
    end
  end

  def remove(data)
    return "List is empty." if @head.nil?

    if @head.data == data
      @head = @head.next
      return "Removed #{data}"
    end

    temp = @head
    while temp.next && temp.next.data != data
      temp = temp.next
    end

    if temp.next
      temp.next = temp.next.next
      return "Removed #{data}"
    end
    "Number not found."
  end

  def display
    return "List is empty." if @head.nil?

    elements = []
    temp = @head
    while temp
      elements << temp.data
      temp = temp.next
    end
    elements
  end

  def contains?(data)
    temp = @head
    while temp
      return true if temp.data == data
      temp = temp.next
    end
    false
  end
end

def menu
  ll = LinkedList.new
  loop do
    puts "\nLinked List Menu:"
    puts "1. Insert"
    puts "2. Remove"
    puts "3. Display List"
    puts "4. Check if Number Exists"
    puts "5. Exit"
    print "Enter choice: "
    choice = gets.chomp.to_i

    case choice
    when 1
      print "Enter number to insert: "
      num = gets.chomp.to_i
      ll.insert(num)
    when 2
      print "Enter number to remove: "
      num = gets.chomp.to_i
      puts ll.remove(num)
    when 3
      puts "List: #{ll.display}"
    when 4
      print "Enter number to check: "
      num = gets.chomp.to_i
      puts ll.contains?(num) ? "Exists" : "Does not exist"
    when 5
      break
    else
      puts "Invalid choice, try again."
    end
  end
end

menu
